source/server/admin/stats_handler.cc

```diff
@@ -225,24 +249,84 @@ StatsHandler::statsAsJson(const std::map<std::string, uint64_t>& all_stats,
 
   ProtobufWkt::Struct histograms_obj_container;
   auto* histograms_obj_container_fields = histograms_obj_container.mutable_fields();
-  std::vector<ProtobufWkt::Value> computed_quantile_array;
 
-  bool found_used_histogram = false;
-  for (const Stats::ParentHistogramSharedPtr& histogram : all_histograms) {
-    if (shouldShowMetric(*histogram, used_only, regex)) {
-      if (!found_used_histogram) {
-        // It is not possible for the supported quantiles to differ across histograms, so it is ok
-        // to send them once.
-        Stats::HistogramStatisticsImpl empty_statistics;
-        std::vector<ProtobufWkt::Value> supported_quantile_array;
-        for (double quantile : empty_statistics.supportedQuantiles()) {
-          supported_quantile_array.push_back(ValueUtil::numberValue(quantile * 100));
-        }
-        (*histograms_obj_fields)["supported_quantiles"] =
-            ValueUtil::listValue(supported_quantile_array);
-        found_used_histogram = true;
+  // Display bucket data if histogram_buckets query parameter is used, otherwise output contains
+  // quantile summary data.
+  if (!histogram_buckets_value.has_value()) {
+    std::vector<ProtobufWkt::Value> computed_quantile_array =
+        statsAsJsonQuantileSummaryHelper(all_histograms, used_only, regex);
+    if (!computed_quantile_array.empty()) { // If used histogram found.
+      // It is not possible for the supported quantiles to differ across histograms, so it is ok
```

Comment:
this function has gotten too long to read -- can you factor out the blocks of this if/elseif chain into three helper methods?

I can't quite see it now but maybe when you do that there'll be a mechanism to factor out some of the duplication between the implementations, e.g. whether statsAsJsonDisjointHistogramBucketsHelper or a different helper is used to get the buckets.

Reply:
I did some refactoring, but statsAsJsonCumulativeHistogramBucketsHelper and statsAsJsonDisjointHistogramBucketsHelper still look very similar.